R/cf.R

```diff
@@ -146,14 +146,17 @@ jackknife_error <- function (samples, na.rm = FALSE) {
   sqrt(factor * (N - 1)^2 / N) * sd(samples)
 }
 
+#' Computes covariance matrix for jackknife samples. The rows containing any NA will be deleted if na.rm = TRUE.
+#' 
+#'
 #' @export
 jackknife_cov <- function (x, y = NULL, na.rm = FALSE, ...) {
     factor <- 1.0
     
     if (is.null(y)) {
         N <- nrow(x)
         if (na.rm) {
-            na_values <- apply(x, 2, function (row) any(is.na(row)))
+            na_values <- apply(x, 1, function (row) any(is.na(row)))
```

Comment:
Does this actually happen in practice using Jackknife samples? Leaving out a single measurement should, I would guess, not really lead to things suddenly becoming badly defined. It's a different matter for the bootstrap, I would say.

Reply:
I am not sure what you are aiming at. Matthias has the problem that there are some `NA` values in the effective masses that Markus gave him.

For the bootstrap leaving out samples is no problem, the distribution is not affected by it, how well it is resolved. For jackknife the number of samples is intrinsically linked to the width of the distribution, therefore omitting samples makes one underestimate the error.

Reply:
> I am not sure what you are aiming at. Matthias has the problem that there are some NA values in the effective masses that Markus gave him.

But why would this even happen in a jackknife sample? My point is that I find it very unlikely that an effective mass works correctly on 251 measurements but then fails to compute correctly on a subset of 250 measurements. Are you sure that these effective masses use jackknife samples?

> For the bootstrap leaving out samples is no problem, the distribution is not affected by it, how well it is resolved. For jackknife the number of samples is intrinsically linked to the width of the distribution, therefore omitting samples makes one underestimate the error.

I know, that's not the origin of my question.

Reply:
Maybe it's not the best place to discuss it here but principally you are right Bartek. The problem for me is, that I have to cut out time slices which are NA in my routine. There are some correlators/effectivemasses which have NA all over one column . If I cut out all rows which have NA then I have no covariance matrix anymore. That's why I ran my routine over all rho data cutting out all time slices which contained any NA in a column. Doing this I need to multiply the respective covariance matrix just by the global factor which is inserted at jackknife_cov(...).

The main problem for my routine is that it is dependent on the method which is applied to the correlators to obtain the effectivemass data. That's why we need to think of a method to handle the used/needed "x"-values (without NAs). I need df <- data.frame(x = x, y = y, dy = dy) and the inverse of the cov matrix which I handle separately to the function call which runs the whole routine. That would be the only "interface" between LuescherAnalysis and BayesianFitting.

I will try to finish the R package over the weekend such that we all can have a look at it and test it if you are interested.

Reply:
NA's in one or multiple **columns** are not a surprise. Hadron does not currently employ periodic boundary conditions in the computation of the correlator time derivative or the effective mass, such that one or multiple columns become undefined in the procedure. These columns should be **completely ignored** in any case and you should know **at the calling point of your routine**, that these columns are not to part of the dataset. In particular, removing **columns** from the bootstrap or jackknife matrix **must not change the jackknife error scaling factor**.

The situation we are discussing with @martin-ueding is a different one: when bootstrapping, the resampled correlator can behave so weirdly that the computation of the effective mass breaks down. Thus, on some bootstrap samples (**rows** of the bootstrap matrix, not columns), the effective mass might be NA for one or a few time slices. In the bootstrap case, the data is then simply resampled unless there are too many of these. In the jackknife case, resampling is not possible, so the treatment of the error and covariance must be adjusted.

I am very surprised that one would find a similar situation on jackknife samples, unless there are only very few measurements in total.

Reply:
> The main problem for my routine is that it is dependent on the method which is applied to the correlators to obtain the effectivemass data. That's why we need to think of a method to handle the used/needed "x"-values (without NAs). I need df <- data.frame(x = x, y = y, dy = dy) and the inverse of the cov matrix which I handle separately to the function call which runs the whole routine. That would be the only "interface" between LuescherAnalysis and BayesianFitting.

Yes that's a good interface.  Thus, at the calling point in your analysis (where you know which data you are passing to BayesianFitting), you should restrict yourself to the x values for which the effective mass of that particular type is well defined. This should certainly not be done inside of ```jackknife_cov```.

Reply:
Yes, you are absolutely right. I just wanted to say that cutting rows is not that important to "my" routine since I need to cut columns first. I just spotted the bug when I looked at `jackknife_cov' because I considered using it. I thought that one could extend this to cut out columns which are completely NA within `jackknife_cov`. But now it is clear to me why we do not want this at this function. However, I should use `jackknife_cov` after removing the NA cols. Thanks.